Add a sample for AuthContext (#45)
Azure-Samples/ms-identity-msal-java-samples: 3. Java Servlet Web App Tutorial/1-Authentication/sign-in-auth-context/README.md

---
page_type: sample
languages:
  - java
products:
  - azure
  - msal-java

>   - azure-active-directory

Reviewer: ```suggestion
  - entra
```

> name: Enable your Java Servlet web app to sign in users with the Microsoft Entra platform and step-up the authentication challenge on demand.

Reviewer: ```suggestion
```

The name is redundant in the presence of a H1.

> urlFragment: ms-identity-java-servlet-webapp-authentication

Reviewer: ```suggestion
urlFragment: java-servlet-webapp-authentication
```

> description: "This sample demonstrates a Java Servlet web app that signs in users with the Microsoft Entra platform and to step up the authentication challenge on demand."

Reviewer: ```suggestion
description: "This sample demonstrates a Java Servlet web app that signs users in with the Microsoft Entra platform and steps up the authentication challenge on-demand."
```

---
# Enable your Java Servlet web app to sign in users to your Azure Active Directory tenant with the Microsoft identity platform

## Overview

> This sample demonstrates a Java Servlet web app that signs in users to your Azure Active Directory tenant using the [Microsoft Authentication Library (MSAL) for Java](https://github.com/AzureAD/microsoft-authentication-library-for-java).

Reviewer: ```suggestion
This sample demonstrates a Java Servlet web app that signs users in to your Microsoft Entra tenant using the [Microsoft Authentication Library (MSAL) for Java](https://learn.microsoft.com/entra/msal/java/).
```

It also demonstrates how to step-up the authentication challenge by requiring multi-factor authentication (MFA) for certain routes.

![Overview](./ReadmeFiles/topology.png)

> Important: Make sure you understand the [basic web site authentication sample](../sign-in) first. This shows how to sing-in users.

Reviewer: ```suggestion
>**Note**
>Make sure you understand the [basic web site authentication sample](../sign-in) first. This shows how to sign users in.
```

## Scenario

This web application shows how to challenge users to authenticate for some routes and challenge users to authenticate and perform MFA for other routes.

> 1. Some web site actions (routes) need the user to be authenticated and some web site actions (routes) require the user to be 

Reviewer: ```suggestion
1. Some web site actions (routes) need the user to be authenticated and others require the user to be
```

> have authenticated and to have passed MFA.

Reviewer: ```suggestion
authenticated and to have passed the MFA challenge.
```

> 1. An authentication context "c1" has already been created and associated with MFA

Reviewer: ```suggestion
1. An authentication context `c1` has already been created and associated with MFA
```

> 1. If an MFA route is accessed by the user, the ID token is checked for proof of MFA - a claim named "arcs" with the value "C1"

Reviewer: ```suggestion
1. If an MFA route is accessed by the user, the ID token is checked for proof of MFA - a claim named `arcs` with the value `C1`
```

1. If an id token does not exist (i.e. user is not authenticated at all) or if the claim does not exist (they have not MFA-ed), the user is challenged to MFA. This means that they are redirected to the authorization page and a claims challenge is appended to the authorization URI. 

Example claim challenge: `{"id_token":{"acrs":{"essential":true,"value":"c1"}}}`

> **Important** MFA is only one of the many [Conditional Access](https://learn.microsoft.com/azure/active-directory/conditional-access/overview) controls available. Any of them may be used.

Reviewer: ```suggestion
>**Note**
>MFA is one of the many [Conditional Access](https://learn.microsoft.com/azure/active-directory/conditional-access/overview) controls available. Any of them may be used.
```

    - You'll need this key later in your code's configuration files. This key value will not be displayed again, and is not retrievable by any other means, so make sure to note it from the Azure portal before navigating to any other screen or blade.

### Configure the web app (java-servlet-webapp-authentication) to use your app registration

Open the project in your IDE to configure the code.

> In the steps below, "ClientID" is the same as "Application ID" or "AppId".

1. Open the `./src/main/resources/authentication.properties` file
2. Find the string `{enter-your-tenant-id-here}`. Replace the existing value with you tenant id
3. Find the string `{enter-your-client-id-here}` and replace the existing value with the application ID (clientId) of the `java-servlet-webapp-authentication` application copied from the Azure portal.
4. Find the string `{enter-your-client-secret-here}` and replace the existing value with the key you saved during the creation of the `java-servlet-webapp-authentication` app, in the Azure portal.

</details>

### 2. Setup the MFA Conditional Access Context

> 1. In the Azure Portal, go to the Conditional Access section.

Reviewer: ```suggestion
1. In the Azure Portal, go to the Microsoft Entra Conditional Access section.
```

    - Select "Target Resources" and set it to "Authentication Context". Then add the context you just created.
    - You can skip the "Conditions" part of the flow. This policy will apply to all calls.
    - In Grant, check the "Require Multi-Factor Authentication" policy.
    - You can skip the "Session" 

![CA-MFA](./ReadmeFiles/CA-MFA.png)

## Running The Sample

#### Build .war File Using Maven

1. Navigate to the directory containing the pom.xml file for this sample (the same directory as this README), and run the following Maven command:
    ```
    mvn clean package
    ```
1. This should generate a `.war` file which can be run on a variety of application servers

#### Deploying the Sample

Our samples can be deployed to a number of application servers, such as Tomcat, WebLogic, or Webshpere, and MSAL Java itself can generally be integrated into existing applications without changes to your existing deployment set up.

You can find instructions for deploying our samples [here on MSAL Java's Github wiki](https://github.com/AzureAD/microsoft-authentication-library-for-java/wiki/Deployment-Instructions-for-MSAL-Java-Samples).

## Explore the sample

> - In `authwebapp` there are 2 protected routes: "token_details" requires simple authentication and "token_details_mfa" which requires MFA.

Reviewer: ```suggestion
- In `authwebapp` there are 2 protected routes: `token_details` requires simple authentication and `token_details_mfa` which requires MFA.
```

> - The orchestration logic is in the AuthenticationFilter. It figures out the state of the user (authenticated, authenticated + MFA, not authenticated) and the requirements of the route (based on configuration).

Reviewer: ```suggestion
- The orchestration logic is in the `AuthenticationFilter`. It figures out the state of the user (authenticated, authenticated + MFA, not authenticated) and the requirements of the route (based on configuration).
```

- If the auth requirement is not met, the user is challenged, meaning they are redirected to the Identity Provider (Microsoft Entra).

> - To understand if a user meets the authentication context (MFA), the sample looks at the ID Token for a claim named "acrs" and value "c1"

Reviewer: ```suggestion
- To understand if a user meets the authentication context (MFA), the sample looks at the ID Token for a claim named `acrs` and value `c1`.
```

- To challenge a user if they are not authenticated, the sample adds a claims challenge to the authorization URI `claims={"id_token":{"acrs":{"essential":true,"value":"c1"}}}`

> :information_source: Did the sample not work for you as expected? Did you encounter issues trying this sample? Then please reach out to us using the [GitHub Issues](../../../../issues) page.

## We'd love your feedback!

> Were we successful in addressing your learning objective? Consider taking a moment to [share your experience with us](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR73pcsbpbxNJuZCMKN0lURpURDQwVUxQWENUMlpLUlA0QzdJNVE3TUJRSyQlQCN0PWcu).

Reviewer: ```suggestion
Were we successful in addressing your learning objective? Consider taking a moment to [share your experience with us](https://github.com/Azure-Samples/ms-identity-msal-java-samples/issues).
```

Let's point to GitHub Issues - not sure who maintains the Form anymore.

## Next Steps or Deploy to Azure

> As next steps, we can now either [get an Access Token for the users we signed-in in this tutorial](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-call-graph), or we can proceed [to deploy this app to the **Azure App Service**](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-authentication/tree/main/4-Deployment/deploy-to-azure-app-service).

Reviewer: ```suggestion
As next steps, we can now either [get an Access Token for the users we signed-in in this tutorial](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-call-graph), or we can proceed [to deploy this app to Azure App Service](https://github.com/Azure-Samples/ms-identity-java-servlet-webapp-authentication/tree/main/4-Deployment/deploy-to-azure-app-service).
```

Ask your questions on Stack Overflow first and browse existing issues to see if someone has asked your question before.
Make sure that your questions or comments are tagged with [`azure-active-directory` `ms-identity` `java` `msal`].

If you find a bug in the sample, please raise the issue on [GitHub Issues](../../../../issues).

To provide a recommendation, visit the following [User Voice page](https://feedback.azure.com/forums/169401-azure-active-directory).

## Contributing

This project welcomes contributions and suggestions. Most contributions require you to agree to a Contributor License Agreement (CLA) declaring that you have the right to, and actually do, grant us the rights to use your contribution. For details, visit https://cla.opensource.microsoft.com

## Code of Conduct

This project has adopted the Microsoft Open Source Code of Conduct. For more information see the Code of Conduct FAQ or contact opencode@microsoft.com with any additional questions or comments.

## More information

- [Microsoft Authentication Library \(MSAL\) for Java](https://github.com/AzureAD/microsoft-authentication-library-for-java)

> - [MSAL Java Reference Documentation](http://javadoc.io/doc/com.microsoft.azure/msal4j)

Reviewer: ```suggestion
- [MSAL Java Reference Documentation](https://learn.microsoft.com/java/api/overview/msal-java/)
```

> - [Microsoft identity platform (Azure Active Directory for developers)](https://docs.microsoft.com/azure/active-directory/develop/)

Reviewer: ```suggestion
- [Microsoft identity platform (Azure Active Directory for developers)](https://learn.microsoft.com/azure/active-directory/develop/)
```

> - [Quickstart: Register an application with the Microsoft identity platform (Preview)](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app)

Reviewer: ```suggestion
- [Quickstart: Register an application with the Microsoft identity platform (Preview)](https://learn.microsoft.com/azure/active-directory/develop/quickstart-register-app)
```

> - [Understanding Azure AD application consent experiences](https://docs.microsoft.com/azure/active-directory/develop/application-consent-experience)

Reviewer: ```suggestion
- [Understanding Azure AD application consent experiences](https://learn.microsoft.com/azure/active-directory/develop/application-consent-experience)
```

> - [Understand user and admin consent](https://docs.microsoft.com/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant#understand-user-and-admin-consent)

Reviewer: ```suggestion
- [Understand user and admin consent](https://learn.microsoft.com/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant#understand-user-and-admin-consent)
```

> - [MSAL code samples](https://docs.microsoft.com/azure/active-directory/develop/sample-v2-code)

Reviewer: ```suggestion
- [MSAL code samples](https://learn.microsoft.com/azure/active-directory/develop/sample-v2-code)
```

---
page_type: sample
languages:
  - java
products:
  - azure
  - msal-java
  - entra
urlFragment: java-servlet-webapp-authentication
description: "This sample demonstrates a Java Servlet web app that signs users in with the Microsoft Entra platform and steps up the authentication challenge on-demand."
---
# Enable your Java Servlet web app to sign in users to your Azure Active Directory tenant with the Microsoft identity platform

## Overview

This sample demonstrates a Java Servlet web app that signs users in to your Microsoft Entra tenant using the [Microsoft Authentication Library (MSAL) for Java](https://learn.microsoft.com/entra/msal/java/).
It also demonstrates how to step-up the authentication challenge by requiring multi-factor authentication (MFA) for certain routes.

![Overview](./ReadmeFiles/topology.png)

>**Note**

> >Make sure you understand the [basic web site authentication sample](https://github.com/Azure-Samples/ms-identity-msal-java-samples/tree/bogavril/auth-context/3-java-servlet-web-app/1-Authentication/sign-in) first. This shows how to sign users in.

Reviewer 2: I updated the link here to fix the broken path

Reviewer 2: I'll make sure my next change follows the proper format. Suggested changes like Den. My bad

4. Find the string `{enter-your-client-secret-here}` and replace the existing value with the key you saved during the creation of the `java-servlet-webapp-authentication` app, in the Azure portal.

</details>

### 2. Setup the MFA Conditional Access Context

1. In the Azure Portal, go to the Microsoft Entra Conditional Access section.

![CA](./ReadmeFiles/CA.png)

Note: this is a different Azure Portal section than Microsoft Entra ID

2. Go to the "Authentication Contexts" page, and create a new context. Give it a name (e.g. "MFA") and select and id from the list, for example "c1". Ensure the "Publish To Apps" checkbox is checked. Make a note of the ID.
3. Now go to Policies and create a new Conditional Access policy. 
    - Select "All Users" or a subset of users
    - Select "Target Resources" and set it to "Authentication Context". Then add the context you just created.
    - You can skip the "Conditions" part of the flow. This policy will apply to all calls.
    - In Grant, check the "Require Multi-Factor Authentication" policy.
    - You can skip the "Session" 

> 

Reviewer: ```suggestion

Make sure that the policy is enabled by selecting **On** when configuring the policy in the wizard.
```

![CA-MFA](./ReadmeFiles/CA-MFA.png)

## Running The Sample

#### Build .war File Using Maven

1. Navigate to the directory containing the pom.xml file for this sample (the same directory as this README), and run the following Maven command:
    ```
    mvn clean package
    ```

> 1. This should generate a `.war` file which can be run on a variety of application servers

Reviewer: ```suggestion
1. If the build is successful, it will generate a `.war` file in the `target` folder which can be run on a variety of application servers.
```

#### Deploying the Sample

Our samples can be deployed to a number of application servers, such as Tomcat, WebLogic, or Webshpere, and MSAL Java itself can generally be integrated into existing applications without changes to your existing deployment set up.

> You can find instructions for deploying our samples [here on MSAL Java's Github wiki](https://github.com/AzureAD/microsoft-authentication-library-for-java/wiki/Deployment-Instructions-for-MSAL-Java-Samples).

Reviewer: ```suggestion
You can find instructions for deploying our samples [in the MSAL4J documentation](https://learn.microsoft.com/entra/msal/java/advanced/deployment-instructions-for-msal-java-samples).
```
